### PR TITLE
Add NSQ as a stream source 

### LINKED ISSUE #64: Add NSQ as a stream source
Add [NSQ](http://nsq.io/) alongside existing Kinesis support.

This is a key building block for:
1. Swapping out the Clojure Collector for the Stream Collector without using Kinesis
2. Using this to archive events in other clouds (e.g. Azure, GCP) - will need further tickets to support Azure/Google Blob Storage
3. Adding writing files to S3 to Snowplow Mini

### PR DESCRIPTION
**Pull request by aldemirenes**

### CODE REVIEW

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+
+/**
+ * @param kinesisConfig the kinesis configuration
+ * @param nsqConfig the NSQ configuration
+ * @param badSink the configured BadSink
+ * @param serializer the instance of one of the serializer
+ * @param maxConnectionTime max time for trying to connect S3 instance 
+ */
+class NSQSourceExecutor(
```
Comment (alexanderdean) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala]: Should be `NsqSourceExecutor`

File: src/main/resources/config.hocon.sample [hunk 2]
```diff
@@ -1,7 +1,13 @@
 # Default configuration for kinesis-lzo-s3-sink
 
+# Sources currently supported are:
+# 'kinesis' for reading records from a Kinesis stream
+# 'NSQ' for reading records from NSQ
+source = "{{sinkKinesisSource}}"
```
Comment (BenFradet) [inline src/main/resources/config.hocon.sample]: you can just have `{{source}}` as default value

File: src/main/resources/config.hocon.sample [hunk 3]
```diff
@@ -1,7 +1,13 @@
...
+
 sink {
 
+
```
Comment (BenFradet) [inline src/main/resources/config.hocon.sample]: no need for an empty line

File: src/main/resources/config.hocon.sample [hunk 4]
```diff
@@ -15,6 +21,16 @@ sink {
     secret-key: "iam"
   }
 
+  # config for NSQ sink and NSQ source
+  NSQ {
+    good-source: "{{sinkKinesisNSQSourceTopicGoodName}}" #topic name for good event source
+    bad-sink: "{{sinkKinesisNSQSinkTopicBadName}}" #topic name for bad event sink
+    nsq-host: "{{NSQHost}}" #host name for NSQ tools
+    nsqd-port: "{{nsqdPort}}" #port for nsqd
+    nsqlookupd-port: {{nsqlookupdPort}} #port for nsqlookupd
```
Comment (BenFradet) [inline src/main/resources/config.hocon.sample]: why do we need 2 ports?

File: src/main/resources/config.hocon.sample [hunk 5]
```diff
@@ -15,6 +21,16 @@ sink {
...
+    buffer-size: {{sinkKinesisNSQBufferSize}} #number of record from NSQ before sending to S3
```
Comment (BenFradet) [inline src/main/resources/config.hocon.sample]: please format your comments like the others in this file, e.g. `# Number of...`

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+  private def sleep(sleepTime: Long): Unit = {
+    try {
+    Thread.sleep(sleepTime)
+    } catch {
+      case e: InterruptedException => ()
+    }
+  }
+
+  /**
+  * Terminate the application in a way the KCL cannot stop
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala]: please refrain from copy/pasting code

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+  /**
+  * Returns an ISO valid timestamp
+  *
+  * @param tstamp The Timestamp to convert
+  * @return the formatted Timestamp
+  */
+  private def getTimestamp(tstamp: Long): String = {
+    val dt = new DateTime(tstamp)
+    TstampFormat.print(dt)
+  }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NSQSourceExecutor.scala]: if there are methods shared between multiple classes, refactor them out

File: src/main/resources/config.hocon.sample [hunk 8]
```diff
@@ -63,9 +89,9 @@ sink {
   # - the number of stored records exceeds record-limit or
   # - the time in milliseconds since it was last emptied exceeds time-limit
   buffer {
-    byte-limit: {{sinkLzoBufferByteThreshold}}
+    byte-limit: {{sinkLzoBufferByteThreshold}} # Not supported by NSQ; will be ignored
     record-limit: {{sinkLzoBufferRecordThreshold}}
-    time-limit: {{sinkLzoBufferTimeThreshold}}
+    time-limit: {{sinkLzoBufferTimeThreshold}} # Not supported by NSQ; will be ignored
```
Comment (BenFradet) [inline src/main/resources/config.hocon.sample]: same thing here, those properties need to be used

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+// Joda-Time
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.DateTimeFormat
+
+// This project
+import sinks._
+import serializers._
+
+
+/**
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala]: missing scaladoc

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    val currentTimeObject = new DateTime(System.currentTimeMillis())
+    val startTimeObject = new DateTime(startTime)
+    val endTimeObject = new DateTime(endTime)
+
+    DateFormat.print(currentTimeObject) + "-" + 
+    TimeFormat.print(startTimeObject)   + "-" + 
+    TimeFormat.print(endTimeObject)
+  }
+  
+  private def convertRawRecord(rawRecord: RawRecord): EmitterInput = rawRecord.success
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala]: I don't think this is really useful

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+              }
+
+              if (failures.size > 0) {
+                s3EmitterUtils.sendFailures(failures)
+              }
+
+              msgBuffer.clear()
+              //make buffer start time of the next buffer the buffer finish time of the current buffer 
+              bufferStartTime = bufferEndTime
+            }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala]: indent

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+            }                
+        }
+        msg.finished() 
+      }
+    }
+
+    val lookup = new DefaultNSQLookup
+    // use NSQLookupd
+    lookup.addLookupAddress(nsqConfig.nsqHost, nsqConfig.nsqlookupdPort)
+    val consumer = new NSQConsumer(lookup, nsqConfig.nsqGoodSourceTopicName, nsqConfig.nsqGoodSourceChannelName, nsqCallback)
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala]: same here, isn't there an error callback?

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+          msgBuffer += validMsg
+            if (msgBuffer.size == nsqBufferSize) {
+              //finish time of filling the buffer
+              val bufferEndTime = System.currentTimeMillis()
+              val baseFilename = getBaseFilename(bufferStartTime, bufferEndTime)
+              val serializationResults = serializer.serialize(msgBuffer.toList, baseFilename)
+              val (successes, failures) = serializationResults.results.partition(_.isSuccess)
+
+              if (successes.size > 0) {
+                serializationResults.namedStreams.foreach { s3EmitterUtils.attemptEmit(_, connectionAttemptStartTime) }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/NsqSourceExecutor.scala]: the result of attemptEmit isn't used

File: src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/S3EmitterUtils.scala [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+            case None => None
+          }
+          attemptCount = attemptCount + 1
+          sleep(BackoffPeriod)
+        }
+      }
+    }
+    false
+  }
+}
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.snowplow.storage.kinesis/s3/S3EmitterUtils.scala]: This whole refactor needs to either have a dedicated issue or not be done at all.

**Pull request closed without merging**